src/cargo/ops/cargo_rustc/context/mod.rs

```diff
@@ -246,11 +246,17 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
     /// all the units mentioned in `units`.
     pub fn probe_target_info(&mut self) -> CargoResult<()> {
         debug!("probe_target_info");
-        self.probe_target_info_kind(Kind::Target)?;
-        if self.requested_target().is_none() {
```

Comment:
Hm, interesting, so the previous logic kinda handled this in the case where `target` was not passed at all....

Perhaps add a helper `Context::is_cross_compiling` which handles both "no target" and "explicit target the same as the host", so that we don't duplicated this logic incorrectly elsewhere?